File: python/src/ocrinfer/diagnostics.py
from __future__ import annotations

from collections.abc import Iterable

import pandas as pd

from .core import align_text


def _policy(x: pd.DataFrame) -> dict:
    return x.attrs.get("ocrinfer_policy") or {}


def confusion_table(
    x: pd.DataFrame,
    *,
    metric: str = "cer",
    by: Iterable[str] | str | None = None,
    include_equal: bool = False,
    epsilon: str = "<eps>",
) -> pd.DataFrame:
    if metric not in {"cer", "wer"}:
        raise ValueError("metric must be cer or wer")
    by_cols = [] if by is None else ([by] if isinstance(by, str) else list(by))
    required = {"metric", "reference", "prediction", *by_cols}
    missing = required.difference(x.columns)
    if missing:
        raise ValueError("missing columns: " + ", ".join(sorted(missing)))

    policy = _policy(x)
    char_unit = policy.get("char_unit", "grapheme")
    whitespace = policy.get("whitespace", "preserve")
    align_ws = "collapse" if metric == "wer" and whitespace == "preserve" else whitespace

    rows = []
    d = x[x["metric"].astype(str) == metric]
    for _, source in d.iterrows():
        alignment = align_text(
            str(source["reference"]),
            str(source["prediction"]),
            unit=char_unit if metric == "cer" else "word",
            unicode=policy.get("unicode", "NFC"),
            case=policy.get("case", "preserve"),
            whitespace=align_ws,
            punctuation=policy.get("punctuation", "preserve"),
        )
        for item in alignment:
            if not include_equal and item.operation == "equal":
                continue
            row = {col: source[col] for col in by_cols}
            row.update(
                {
                    "reference_token": epsilon if item.reference is None else item.reference,
                    "prediction_token": epsilon if item.hypothesis is None else item.hypothesis,
                    "operation": item.operation,
                }
            )
            rows.append(row)

    if not rows:
        return pd.DataFrame(
            columns=[*by_cols, "reference_token", "prediction_token", "operation", "n"]
        )

    raw = pd.DataFrame(rows)
    group_cols = [*by_cols, "reference_token", "prediction_token", "operation"]
    out = (
        raw.groupby(group_cols, dropna=False, sort=False)
        .size()
        .reset_index(name="n")
        .sort_values("n", ascending=False, kind="stable")
        .reset_index(drop=True)
    )
    return out


def error_profile(
    x: pd.DataFrame,
    *,
    by: Iterable[str] | str | None = None,
) -> pd.DataFrame:
    by_cols = [] if by is None else ([by] if isinstance(by, str) else list(by))
    required = {
        "metric", "n_reference", "substitutions",
        "deletions", "insertions", "distance", *by_cols,
    }
    missing = required.difference(x.columns)
    if missing:
        raise ValueError("missing columns: " + ", ".join(sorted(missing)))

    group_cols = [*by_cols, "metric"]
    rows = []
    for keys, d in x.groupby(group_cols, dropna=False, sort=False):
        if not isinstance(keys, tuple):
            keys = (keys,)
        meta = dict(zip(group_cols, keys))
        ref = int(d["n_reference"].sum())
        sub = int(d["substitutions"].sum())
        delete = int(d["deletions"].sum())
        insert = int(d["insertions"].sum())
        dist = int(d["distance"].sum())

        def rate(value: int) -> float:
            return float("nan") if ref == 0 else value / ref

        rows.append(
            {
                **meta,
                "n_rows": len(d),
                "n_reference": ref,
                "substitutions": sub,
                "deletions": delete,
                "insertions": insert,
                "distance": dist,
                "substitution_rate": rate(sub),
                "deletion_rate": rate(delete),
                "insertion_rate": rate(insert),
                "error_rate": (
                    0.0 if ref == 0 and insert == 0
                    else float(insert) if ref == 0
                    else dist / ref
                ),
            }
        )

    out = pd.DataFrame(rows)
    out.attrs["ocrinfer_policy"] = x.attrs.get("ocrinfer_policy")
    return out


def extract_error_spans(
    x: pd.DataFrame,
    *,
    metric: str = "cer",
    by: Iterable[str] | str | None = None,
    context: int = 5,
) -> pd.DataFrame:
    if metric not in {"cer", "wer"}:
        raise ValueError("metric must be cer or wer")
    if context < 0:
        raise ValueError("context must be non-negative")

    by_cols = [] if by is None else ([by] if isinstance(by, str) else list(by))
    required = {"metric", "reference", "prediction", "input_row", *by_cols}
    missing = required.difference(x.columns)
    if missing:
        raise ValueError("missing columns: " + ", ".join(sorted(missing)))

    policy = _policy(x)
    char_unit = policy.get("char_unit", "grapheme")
    whitespace = policy.get("whitespace", "preserve")
    align_ws = "collapse" if metric == "wer" and whitespace == "preserve" else whitespace
    sep = " " if metric == "wer" else ""

    def collapse(tokens):
        return sep.join(token for token in tokens if token is not None)

    rows = []
    d = x[x["metric"].astype(str) == metric]

    for _, source in d.iterrows():
        alignment = align_text(
            str(source["reference"]),
            str(source["prediction"]),
            unit=char_unit if metric == "cer" else "word",
            unicode=policy.get("unicode", "NFC"),
            case=policy.get("case", "preserve"),
            whitespace=align_ws,
            punctuation=policy.get("punctuation", "preserve"),
        )
        if not alignment or all(a.operation == "equal" for a in alignment):
            continue

        ref_pos = []
        hyp_pos = []
        r = h = 0
        for a in alignment:
            if a.reference is not None:
                r += 1
            if a.hypothesis is not None:
                h += 1
            ref_pos.append(r)
            hyp_pos.append(h)

        error_indices = [i for i, a in enumerate(alignment) if a.operation != "equal"]
        runs = []
        start = prev = error_indices[0]
        for idx in error_indices[1:]:
            if idx == prev + 1:
                prev = idx
            else:
                runs.append((start, prev))
                start = prev = idx
        runs.append((start, prev))

        for span_id, (start, end) in enumerate(runs, start=1):
            part = alignment[start:end + 1]
            ops = [a.operation for a in part]
            operation = ops[0] if len(set(ops)) == 1 else "mixed"

            ref_positions = [
                ref_pos[i] for i in range(start, end + 1)
                if alignment[i].reference is not None
            ]
            hyp_positions = [
                hyp_pos[i] for i in range(start, end + 1)
                if alignment[i].hypothesis is not None
            ]

            inserted = [a.hypothesis for a in part if a.hypothesis is not None]
            possible_repetition = False
            if ops and all(op == "insertion" for op in ops) and inserted:
                k = len(inserted)
                before = [
                    a.hypothesis for a in alignment[:start]
                    if a.hypothesis is not None
                ]
                after = [
                    a.hypothesis for a in alignment[end + 1:]
                    if a.hypothesis is not None
                ]
                possible_repetition = (
                    (len(before) >= k and before[-k:] == inserted)
                    or (len(after) >= k and after[:k] == inserted)
                )

            before_tokens = [
                (a.reference if a.reference is not None else a.hypothesis)
                for a in alignment[max(0, start - context):start]
            ]
            after_tokens = [
                (a.reference if a.reference is not None else a.hypothesis)
                for a in alignment[end + 1:end + 1 + context]
            ]

            row = {col: source[col] for col in by_cols}
            row.update(
                {
                    "input_row": int(source["input_row"]),
                    "span_id": span_id,
                    "operation": operation,
                    "operations": "+".join(ops),
                    "reference_span": collapse([a.reference for a in part]),
                    "prediction_span": collapse([a.hypothesis for a in part]),
                    "reference_start": min(ref_positions) if ref_positions else pd.NA,
                    "reference_end": max(ref_positions) if ref_positions else pd.NA,
                    "prediction_start": min(hyp_positions) if hyp_positions else pd.NA,
                    "prediction_end": max(hyp_positions) if hyp_positions else pd.NA,
                    "substitutions": ops.count("substitution"),
                    "deletions": ops.count("deletion"),
                    "insertions": ops.count("insertion"),
                    "context_before": collapse(before_tokens),
                    "context_after": collapse(after_tokens),
                    "possible_repetition": possible_repetition,
                    "reference": str(source["reference"]),
                    "prediction": str(source["prediction"]),
                }
            )
            rows.append(row)

    out = pd.DataFrame(rows)
    out.attrs["ocrinfer_policy"] = x.attrs.get("ocrinfer_policy")
    return out

File: R/diagnostics.R
#' Build a token-level confusion table
#'
#' Reconstructs alignments from evaluation results and counts token-to-token
#' substitutions, deletions, and insertions under the same evaluation policy.
#'
#' @param x Evaluation output from [evaluate_recognition()] with preserved text.
#' @param metric Either "cer" or "wer".
#' @param by Optional character vector of metadata columns for stratified
#'   confusion tables (for example "system" or "script_type").
#' @param include_equal Logical. Include correctly matched tokens if TRUE.
#' @param epsilon Label used for the missing side of insertions/deletions.
#'
#' @return A data frame of aligned token pairs, operation type, and counts.
#' @export
confusion_table <- function(
  x,
  metric = c("cer", "wer"),
  by = NULL,
  include_equal = FALSE,
  epsilon = "<eps>"
) {
  metric <- match.arg(metric)

  required <- c("metric", "reference", "prediction")
  missing_required <- setdiff(required, names(x))
  if (length(missing_required) > 0L) {
    stop(
      sprintf(
        "Missing required columns: %s. Re-run evaluate_recognition() with keep_text = TRUE.",
        paste(missing_required, collapse = ", ")
      ),
      call. = FALSE
    )
  }

  if (!is.null(by)) {
    if (!is.character(by)) {
      stop("`by` must be a character vector.", call. = FALSE)
    }
    missing_by <- setdiff(by, names(x))
    if (length(missing_by) > 0L) {
      stop(
        sprintf("Unknown grouping columns: %s.", paste(missing_by, collapse = ", ")),
        call. = FALSE
      )
    }
  }

  if (!is.logical(include_equal) || length(include_equal) != 1L || is.na(include_equal)) {
    stop("`include_equal` must be TRUE or FALSE.", call. = FALSE)
  }

  if (!is.character(epsilon) || length(epsilon) != 1L || is.na(epsilon)) {
    stop("`epsilon` must be one non-missing string.", call. = FALSE)
  }

  d <- x[as.character(x$metric) == metric, , drop = FALSE]
  if (nrow(d) == 0L) {
    out <- data.frame(
      reference_token = character(),
      prediction_token = character(),
      operation = character(),
      n = integer(),
      stringsAsFactors = FALSE
    )
    return(out)
  }

  policy <- attr(x, "ocrinfer_policy")
  if (is.null(policy)) policy <- list()

  char_unit <- if (!is.null(policy$char_unit)) policy$char_unit else "grapheme"
  unicode <- if (!is.null(policy$unicode)) policy$unicode else "NFC"
  case <- if (!is.null(policy$case)) policy$case else "preserve"
  whitespace <- if (!is.null(policy$whitespace)) policy$whitespace else "preserve"
  punctuation <- if (!is.null(policy$punctuation)) policy$punctuation else "preserve"

  align_unit <- if (metric == "cer") char_unit else "word"
  align_whitespace <- if (metric == "wer" && whitespace == "preserve") {
    "collapse"
  } else {
    whitespace
  }

  rows <- lapply(seq_len(nrow(d)), function(i) {
    a <- align_text(
      d$reference[[i]],
      d$prediction[[i]],
      unit = align_unit,
      unicode = unicode,
      case = case,
      whitespace = align_whitespace,
      punctuation = punctuation
    )

    if (!isTRUE(include_equal)) {
      a <- a[a$operation != "equal", , drop = FALSE]
    }

    if (nrow(a) == 0L) return(NULL)

    a$reference_token <- ifelse(
      is.na(a$reference),
      epsilon,
      as.character(a$reference)
    )
    a$prediction_token <- ifelse(
      is.na(a$hypothesis),
      epsilon,
      as.character(a$hypothesis)
    )

    keep_cols <- c("reference_token", "prediction_token", "operation")
    a <- a[, keep_cols, drop = FALSE]

    if (length(by) > 0L) {
      meta <- d[rep(i, nrow(a)), by, drop = FALSE]
      a <- cbind(meta, a)
    }

    a
  })

  rows <- Filter(Negate(is.null), rows)
  if (length(rows) == 0L) {
    out <- data.frame(
      reference_token = character(),
      prediction_token = character(),
      operation = character(),
      n = integer(),
      stringsAsFactors = FALSE
    )
    if (length(by) > 0L) {
      out <- cbind(d[0, by, drop = FALSE], out)
    }
    return(out)
  }

  aligned <- do.call(rbind, rows)
  rownames(aligned) <- NULL

  group_cols <- c(by, "reference_token", "prediction_token", "operation")
  aligned$.one <- 1L

  out <- stats::aggregate(
    aligned$.one,
    by = aligned[group_cols],
    FUN = sum
  )
  names(out)[names(out) == "x"] <- "n"

  out <- out[order(-out$n), , drop = FALSE]
  rownames(out) <- NULL
  out
}

#' Summarise recognition error types
#'
#' Produces micro-aggregated substitution, deletion, insertion, and total error
#' rates, optionally stratified by system or metadata.
#'
#' @param x Evaluation output from [evaluate_recognition()].
#' @param by Optional character vector of grouping columns.
#'
#' @return A data frame with counts and error-type rates.
#' @export
error_profile <- function(x, by = NULL) {
  required <- c(
    "metric", "n_reference", "substitutions",
    "deletions", "insertions", "distance"
  )
  missing_required <- setdiff(required, names(x))
  if (length(missing_required) > 0L) {
    stop(
      sprintf(
        "Missing required evaluation columns: %s.",
        paste(missing_required, collapse = ", ")
      ),
      call. = FALSE
    )
  }

  if (!is.null(by)) {
    if (!is.character(by)) {
      stop("`by` must be a character vector.", call. = FALSE)
    }
    missing_by <- setdiff(by, names(x))
    if (length(missing_by) > 0L) {
      stop(
        sprintf("Unknown grouping columns: %s.", paste(missing_by, collapse = ", ")),
        call. = FALSE
      )
    }
  }

  group_cols <- unique(c(by, "metric"))

  if (nrow(x) == 0L) {
    out <- x[0, group_cols, drop = FALSE]
    out$n_rows <- integer()
    out$n_reference <- numeric()
    out$substitutions <- numeric()
    out$deletions <- numeric()
    out$insertions <- numeric()
    out$distance <- numeric()
    out$substitution_rate <- numeric()
    out$deletion_rate <- numeric()
    out$insertion_rate <- numeric()
    out$error_rate <- numeric()
    return(out)
  }

  group_factor <- do.call(
    interaction,
    c(x[group_cols], list(drop = TRUE, lex.order = TRUE))
  )
  grouped_rows <- split(seq_len(nrow(x)), group_factor)

  rows <- lapply(grouped_rows, function(idx) {
    d <- x[idx, , drop = FALSE]
    ref <- sum(d$n_reference)
    sub <- sum(d$substitutions)
    del <- sum(d$deletions)
    ins <- sum(d$insertions)
    dist <- sum(d$distance)

    denominator_rate <- function(num) {
      if (ref == 0) NA_real_ else num / ref
    }

    cbind(
      d[1, group_cols, drop = FALSE],
      data.frame(
        n_rows = nrow(d),
        n_reference = ref,
        substitutions = sub,
        deletions = del,
        insertions = ins,
        distance = dist,
        substitution_rate = denominator_rate(sub),
        deletion_rate = denominator_rate(del),
        insertion_rate = denominator_rate(ins),
        error_rate = if (ref == 0) {
          if (ins == 0) 0 else ins
        } else {
          dist / ref
        },
        stringsAsFactors = FALSE
      )
    )
  })

  out <- do.call(rbind, rows)
  rownames(out) <- NULL
  attr(out, "ocrinfer_policy") <- attr(x, "ocrinfer_policy")
  out
}
